### DISCUSSION TITLE
Callback state validation

### DISCUSSION THREAD

**Discussion by madhubabu5** [serverlessworkflow/specification#829] [Q&A] [answered]
I'm experiencing a **syntax error** in the **Serverless Workflow Web Editor** and the **Java SDK** while using the **callback state** example provided in the specification. Could you please verify if this issue lies within the specification itself ?

**Comment by ricardozanini** [comment]
@madhubabu5 we have the fix for the web editor: https://github.com/serverlessworkflow/vscode-extension/pull/26

@JBBianchi do you know if this PR will also fix the web editor? I guess we have to open a PR in the website too including your changes?

For the Java SDK, which version are you using?

**Reply by madhubabu5** [reply]
@ricardozanini Thanks for the response. I am using **Java SDK 4.0.3.Final**. Is that fix has to be done in the Java SDK also ?

**Reply by ricardozanini** [reply]
@madhubabu5 This Java SDK version supports spec v0.8, not the one in the main branch. Reading your reply below, I believe that the problem lies there. We accept contributions to the Java SDK, thou. You can implement it in the main branch there and I release a CR1 version of the SDK supporting what you need. wdyt?

**Reply by madhubabu5** [reply]
@ricardozanini Sure. I'll start working on implementing the necessary changes in the main branch of the Java SDK. 

Can I create an issue in the repository for the problem we discussed?

In the meantime, could you provide any specific guidelines or requirements for the implementation ?

**Reply by madhubabu5** [reply, marked as answer]
@ricardozanini @JBBianchi 

Bug:   **!** is missing in the if condition in the version **4.0.3.Final**

```
                if (haveFunctionDefinition(
                      callbackState.getAction().getFunctionRef().getRefName(), functions)) {
                    addValidationError(
                        "CallbackState action function ref does not reference a defined workflow function definition",
                        ValidationError.WORKFLOW_VALIDATION);
                  }
```
This was already fixed and available from version **4.0.5.Final**. So I will upgrade my version. Thanks for your help.

**Comment by JBBianchi** [comment]
@ricardozanini the VSCoode extension isn't the web editor. Both the extension and the web editor are based on the [TypeScript SDK](https://github.com/serverlessworkflow/sdk-typescript). It's probably where the problem lies.

@madhubabu5 could you please provide more info about the said problem, for example a sample workflow reproducing this behavior?

**Reply by madhubabu5** [reply]
@JBBianchi  To replicate the issue described, follow this link [https://github.com/serverlessworkflow/specification/tree/main/examples#Perform-Customer-Credit-Check-Example](url) , below the workflow diagram you can find the complete JSON or YML. When using this  DSL in the Web Editor, an error occurred saying that the **syntax error in graph**. 

In Java SDK , in the functions **isValid()** and **validate()** in  **WorkflowValidator** class giving incorrect output for the above **DSL**.

When  executing the following code : 
**List<ValidationError> validationErrorList = dslWorkflowValidator.setWorkflow(dslWorkflow).validate();**

Even though syntax in the above DSL is correct, I am getting the following validation error : 
 **com.phenom.sh.workflows.common.exceptions.impl.InvalidDSLFileException: CallbackState action function ref does not reference a defined workflow function definition** But I am sure everything is correct.

Can you please check and let me know if I miss anything.

**Reply by JBBianchi** [reply]
Ok, I see. For the web editor part, it's related to a problem in the TS SDK that's been fixed but not deployed to the website yet. State names with a dash would generate an invalid Mermaid Graph. See [issue 183](https://github.com/serverlessworkflow/sdk-typescript/issues/183) and [PR 184](https://github.com/serverlessworkflow/sdk-typescript/pull/184)

So the synthax of the [Perform Customer Credit Check](https://github.com/serverlessworkflow/specification/blob/9f79f8fba04a1aba749104c8a391d9d646cf169b/examples/perform-customer-credit-check.json) workflow is correct but there is indeed a DSL error in the sample that needs to be fixed: the state `check-credit` have an action that references a function called `call-credit-check-microservice` but there is no function with that name. So either the function named `credit-check-function` or the reference to `call-credit-check-microservice` should be renamed to match one an other.

Here is a version of the workflow should work both on the editor and in the Java SDK:
```json
{
    "id": "customercreditcheck",
    "version": "1.0.0",
    "specVersion": "0.8",
    "name": "customercreditcheck",
    "description": "Perform Customer Credit Check",
    "start": "checkcredit",
    "functions": [
        {
            "name": "call-credit-check-microservice",
            "operation": "http://myapis.org/creditcheckapi.json#doCreditCheck"
        },
        {
            "name": "send-rejection-email-function",
            "operation": "http://myapis.org/creditcheckapi.json#rejectionEmail"
        }
    ],
    "events": [
        {
            "name": "credit-check-completed-event",
            "type": "creditCheckCompleteType",
            "source": "creditCheckSource",
            "correlation": [
                {
                    "contextAttributeName": "customerId"
                }
            ]
        }
    ],
    "states": [
        {
            "name": "checkcredit",
            "type": "callback",
            "action": {
                "name": "check-credit",
                "functionRef": {
                    "refName": "call-credit-check-microservice",
                    "arguments": {
                        "customer": "${ .customer }"
                    }
                }
            },
            "eventRef": "credit-check-completed-event",
            "timeouts": {
                "stateExecTimeout": "PT15M"
            },
            "transition": "evaluatedecision"
        },
        {
            "name": "evaluatedecision",
            "type": "switch",
            "dataConditions": [
                {
                    "condition": "${ .creditCheck | .decision == \"Approved\" }",
                    "transition": "startapplication",
                    "name": "startapplication"
                },
                {
                    "condition": "${ .creditCheck | .decision == \"Denied\" }",
                    "transition": "rejectapplication",
                    "name": "reject-application"
                }
            ],
            "defaultCondition": {
                "transition": "rejectapplication"
            }
        },
        {
            "name": "startapplication",
            "type": "operation",
            "actions": [
                {
                    "name": "start-application",
                    "subFlowRef": "startApplicationWorkflowId"
                }
            ],
            "end": true
        },
        {
            "name": "rejectapplication",
            "type": "operation",
            "actionMode": "sequential",
            "actions": [
                {
                    "name": "reject-application",
                    "functionRef": {
                        "refName": "send-rejection-email-function",
                        "arguments": {
                            "applicant": "${ .customer }"
                        }
                    }
                }
            ],
            "end": true
        }
    ]
}
```

**Reply by madhubabu5** [reply]
@JBBianchi The above dsl is working fine in the WebEditor. But on Java SDK , still showing error.

**Reply by JBBianchi** [reply]
@madhubabu5 is it still the same error? I don't have any experience with the Java SDK so I just fixed the error you mentioned, maybe there is an other one down the line.

**Reply by madhubabu5** [reply]
@JBBianchi yes, still getting the same error in Java SDK. Not sure exactly what was that issue, I will check on that.

**Comment by cdavernas** [comment]
Closing, as the discussion was answered and the issue is now stale given it applies to a discontinued version of the spec!
